File: Programmers_Java/Programmers_Java/src/lv2_우박수열정적분.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class lv2_우박수열정적분 {
    public static void main(String[] args) {
        int k = 5;
        int[][] ranges = {{0,0}, {0,-1}, {2,-3}, {3,-3}};
        System.out.println(Arrays.toString(solution(k, ranges)));
    }

    public static List<Integer> makeSequence(int k) {
        List<Integer> seq = new ArrayList<>();
        while (k != 1) {
            seq.add(k);
            k = k%2==0 ? k/2 : k*3+1;
        }
        seq.add(k);

        return seq;
    }

    public static double[] solution(int k, int[][] ranges) {
        double[] answer = new double[ranges.length];

        List<Integer> seq = makeSequence(k);
        for (int i =0; i<ranges.length; i++) {
            if (ranges[i][0] > ranges[i][1] + seq.size() - 1) {
                answer[i] = -1;
                continue;
            } else if (ranges[i][0] == ranges[i][1] + seq.size() - 1) {
                answer[i] = 0;
                continue;
            }

            double total=0;
            for (int j = ranges[i][0]; j< ranges[i][1] + seq.size() - 1; j++) {
                total += (seq.get(j) + seq.get(j+1)) / 2.0;
            }
            answer[i] = total;
        }

        return answer;
    }
}

File: Programmers_Python/lv2_우박수열 정적분.py

def sequence(k):
    seq = []
    while k!=1:
        seq.append(k)
        k = k/2 if k%2==0 else k*3+1
    seq.append(k)
    return seq

def solution(k, ranges):
    ans = []
    seq = sequence(k)
    n = len(seq)
    for a,b in ranges:
        total = 0
        tmp_arr = seq[a:n+b]
        if a >= n+b:
            ans.append(-1)
            continue
        for i in range(len(tmp_arr)-1):     # 사다리꼴 넓이구하기(높이1생략)
            total += (tmp_arr[i]+tmp_arr[i+1])/2
        ans.append(total)
    return ans




k = 5
ranges = [[0,0],[0,-1],[2,-3],[3,-3]]
print(solution(k, ranges))